lib/gooddata/lcm/actions/synchronize_user_filters.rb

```diff
@@ -89,32 +90,34 @@ def call(params)
             dry_run: false
           }
 
-          puts "Synchronizing in mode \"#{mode}\""
+          semaphore = Mutex.new
+
+          params.gdc_logger.info("Synchronizing in mode \"#{mode}\"")
           case mode
           when 'sync_project'
             CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
               filters << row
             end
             filters_to_load = GoodData::UserFilterBuilder.get_filters(filters, symbolized_config)
-            puts "Synchronizing #{filters_to_load.count} filters"
+            params.gdc_logger.info("Synchronizing #{filters_to_load.count} filters")
             project.add_data_permissions(filters_to_load, run_params)
           when 'sync_one_project_based_on_pid'
             CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
               filters << row if row[multiple_projects_column] == project.pid
             end
             filters_to_load = GoodData::UserFilterBuilder.get_filters(filters, symbolized_config)
-            puts "Synchronizing #{filters_to_load.count} filters"
+            params.gdc_logger.info("Synchronizing #{filters_to_load.count} filters")
             project.add_data_permissions(filters_to_load, run_params)
           when 'sync_multiple_projects_based_on_pid'
             CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
               filters << row.to_hash
             end
-            filters.group_by { |u| u[multiple_projects_column] }.flat_map do |project_id, new_filters|
+            filters.group_by { |filter| filter[multiple_projects_column] }.flat_pmap do |project_id, new_filters|
               fail "Project id cannot be empty" if project_id.blank?
-              project = client.projects(project_id)
-              filters_to_load = GoodData::UserFilterBuilder.get_filters(new_filters, symbolized_config)
-              puts "Synchronizing #{filters_to_load.count} filters in project #{project.pid}"
-              project.add_data_permissions(filters_to_load, run_params)
+              client_project = client.projects(project_id)
+              Thread.current['filters_to_load'] = GoodData::UserFilterBuilder.get_filters(new_filters, symbolized_config)
```

Comment:
It took lots of time to detect that if we don't store variable value in `Thread.current` but in `filters_to_load`, we will get the wrong value and I have no idea why.

Reply:
It seems that the reassignment takes place [here](https://github.com/gooddata/gooddata-ruby/blob/master/lib/gooddata/models/user_filters/user_filter_builder.rb#L377) inside `add_data_permissions`. A solution to this would be `dup`ing the `filters_to_load` before passing it to `add_data_permissions`.